File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.cpp
class Solution
{
public:
    vector<vector<int>> combinationSum2(vector<int> &candidates, int target)
    {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vector<int> curr;
        dfs(res, 0, candidates, target, curr);
        return res;
    }

    void dfs(vector<vector<int>> &res, int index, vector<int> &candidates, int target, vector<int> &curr)
    {
        if (target == 0)
        {
            res.push_back(curr);
            return;
        }
        if (target < 0 || index == candidates.size())
        {
            return;
        }
        curr.push_back(candidates[index]);
        dfs(res, index + 1, candidates, target - candidates[index], curr);
        curr.pop_back();
        if (index != candidates.size() - 1 && candidates[index + 1] == candidates[index])
        {
            int current = candidates[index];
            while (index != candidates.size() && candidates[index] == current)
            {
                index++;
            }
            dfs(res, index, candidates, target, curr);
        }
        else
        {
            dfs(res, index + 1, candidates, target, curr);
        }
    }
};

File: Data Structures & Algorithms/search-for-word-ii/submission-1.py
class Node:
    def __init__(self):
        self.children = {}
        self.end = False
        self.index = -1

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        root = Node()
        for i in range(len(words)):
            word = words[i]
            curr = root
            for char in word:
                if char not in curr.children:
                    curr.children[char] = Node()
                curr = curr.children[char]
            curr.end = True
            curr.index = i
        
        res = []
        curr = root

        visited = [[False] * len(board[0]) for _ in range(len(board))]

        def dfs(x, y, node):
            if node.end and node.index != -1:
                res.append(words[node.index])
                node.index = -1

            directions = [[0, 1], [1, 0], [0, -1], [-1, 0]]
            for direction in directions:
                new_x = direction[0] + x
                new_y = direction[1] + y
                if new_x < 0 or new_y < 0 or new_x >= len(board) or new_y >= len(board[0]):
                    continue
                if not visited[new_x][new_y] and board[new_x][new_y] in node.children:
                    visited[new_x][new_y] = True
                    dfs(new_x, new_y, node.children[board[new_x][new_y]])
                    visited[new_x][new_y] = False
            

        for i in range(len(board)):
            for j in range(len(board[0])):
                char = board[i][j]
                if char in root.children:
                    visited[i][j] = True
                    dfs(i, j, root.children[char])
                    visited[i][j] = False
        return res